firedrake/interpolation.py

```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
               interpolation operator.
+        :kwarg default_missing_val: For interpolation across meshes: the
+            optional value to assign to DOFs in the target mesh that are
+            outside the source mesh. If this is not set and an ``output``
+            :class:`.Function` is specified, then such DOF values are left
+            unchanged. If this is not set and no ``output`` :class:`.Function`
+            is specified, then the default value is zero. This does not effect
+            transpose interpolation.
+
         :returns: The resulting interpolated :class:`.Function`.
         """
+        return self.interpolate(*args, **kwargs)
+
+
+class CrossMeshInterpolator(Interpolator):
+    """
+    Interpolate a function from one mesh and function space to another.
+
+    Only works when the function space we interpolate into has point evaluation
+    nodes. A more general version would require finding the global coordinates
+    of all quadrature points of the target function space in the source mesh.
+
+    For arguments, see :class:`.Interpolator`.
+    """
+
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_points=False,
+    ):
+        if subset:
+            raise NotImplementedError("subset not implemented")
+        if freeze_expr:
+            raise NotImplementedError("freeze_expr not implemented")
+        if access != op2.WRITE:
+            raise NotImplementedError("access other than op2.WRITE not implemented")
+        if bcs:
+            raise NotImplementedError("bcs not implemented")
+        if V.ufl_element().mapping() != "identity":
+            # Identity mapping between reference cell and physical coordinates
+            # implies point evaluation nodes
+            raise NotImplementedError(
+                "Can only interpolate into spaces with point evaluation nodes"
+            )
+
+        super().__init__(
+            expr, V, subset, freeze_expr, access, bcs, allow_missing_points
+        )
+
+        self.arguments = extract_arguments(expr)
+        self.nargs = len(self.arguments)
+        if self.nargs and self.arguments[0] != self.expr:
+            raise NotImplementedError(
+                "Can't yet create an interpolator from an expression with arguments."
+            )
+
+        if self._allow_missing_points:
+            missing_points_behaviour = None
+        else:
+            missing_points_behaviour = "error"
+
+        # setup
+        V_dest = V
+        m_src = extract_unique_domain(expr)
+        m_dest = V_dest.ufl_domain()
+        m_src_gdim = m_src.geometric_dimension()
+        m_dest_gdim = m_dest.geometric_dimension()
+        if m_src_gdim != m_dest_gdim:
+            raise ValueError(
+                "geometric dimensions of source and destination meshes must match"
+            )
+        self.m_src = m_src
```

Comment:
Very picky but I don't like this variable name. `src_mesh` for example would make it much clearer what this is.

Reply:
Changed